### PR TITLE
[dagster-airbyte] Generate airbyte assets from API, using cached assets

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Rework of #9702, using the new cached assets API that Owen implemented in #9761 

Adds a new `@experimental` API similar to #9664 that allows users to load Airbyte assets from an Airbyte instance, connecting at initialization-time and pulling connection definitions. The initialization-time configuration is cached, so it's not re-fetched in future repo loads.

```python
airbyte_instance = airbyte_resource.configured(
    {
        "host": "localhost",
        "port": "8000",
        "forward_logs": False,
    }
)
airbyte_assets = load_assets_from_airbyte_instance(
    airbyte_instance,
)
```

Has the same config options as the 'from project' API.

## Test Plan

Tested locally; new unit tests

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #9761** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9761" target="_blank">Graphite</a> 
    * **PR #9766** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9766" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/9766?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/assets_lazy.py
Comment (benpankow) [inline python_modules/dagster/dagster/_core/definitions/assets_lazy.py:16]: list `AssetKeys` are unhashable

File: python_modules/dagster/dagster/_core/definitions/assets_lazy.py
Comment (benpankow) [inline python_modules/dagster/dagster/_core/definitions/assets_lazy.py:19]: misc metadata not assoc. with a particular key

File: python_modules/dagster/dagster/_core/definitions/assets_lazy.py
Comment (benpankow) [inline python_modules/dagster/dagster/_core/definitions/assets_lazy.py:56]: just needed so we can save it as a snapshot, since we can't pass a list directly

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (benpankow) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:63]: this feels a little bad to pass, since `multi_assets` just takes a single group name

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (OwenKephart) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:371]: this unique_id should probably be a hash of some of these input properties (It's definitely pretty unlikely to want to get assets from multiple airbyte instances in a single repository, but best to just avoid collisions in general)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (OwenKephart) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py]: compute_cacheable_data

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (OwenKephart) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:389]: not a huge deal, but you could probably avoid these casts with `check.inst(..., dict)` (not 100% sure)

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (OwenKephart) [inline python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:417]: nit: maybe asset_data_for_connection?

**Review by OwenKephart — COMMENTED** [review]
going to hold off on giving an explicit "approve" until the upstream PR gets merged, but this is looking pretty good

File: python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py
Comment (benpankow) [reply python_modules/libraries/dagster-airbyte/dagster_airbyte/asset_defs.py:397]: Looks like mypy isn't satisfied with this. Will leave for now even though it's a little gross.

**Review by OwenKephart — APPROVED** [review]
awesome

**Pull request merged**
